client/scripts/stores/RecentListingStore.ts

```diff
@@ -0,0 +1,206 @@
+import moment from 'moment';
+import IdStore from './IdStore';
+import { OffchainThread } from '../models';
+import { orderDiscussionsbyLastComment } from '../views/pages/discussions/helpers';
+
+interface IListingParams {
+  topicName?: string;
+  stageName?: string;
+  pinned?: boolean;
+}
+
+interface IListingFetchState {
+  allThreadsInitialized: boolean;
+  topicInitialized: { [topicName: string]: boolean };
+  stageInitialized: { [stageName: string]: boolean };
+
+  allThreadsCutoffDate: moment.Moment;
+  topicCutoffDate: { [topicName: string]: moment.Moment };
+  stageCutoffDate: { [stageName: string]: moment.Moment };
+
+  allThreadsDepleted: boolean;
+  topicDepleted: { [topicName: string]: boolean };
+  stageDepleted: { [stageName: string]: boolean };
+}
+
+class RecentListingStore extends IdStore<OffchainThread> {
+  private _threads = new Array<OffchainThread>();
+  public get threads() {
+    return this._threads;
+  }
+
+  private _fetchState: IListingFetchState = {
+    allThreadsInitialized: false,
+    topicInitialized: {},
+    stageInitialized: {},
+
+    allThreadsCutoffDate: moment(),
+    topicCutoffDate: {},
+    stageCutoffDate: {},
+
+    allThreadsDepleted: false,
+    topicDepleted: {},
+    stageDepleted: {},
+  };
+
+  public add(thread: OffchainThread) {
+    const existingThread = this.getById(thread.id);
+    if (existingThread) this.remove(existingThread);
+
+    super.add(thread);
+    this._threads.push(thread);
+    return this;
+  }
+
+  public remove(thread: OffchainThread) {
+    const existingThread = this.getById(thread.id);
+    if (!existingThread) return;
+    const proposalIndex = this._threads.indexOf(existingThread);
+    if (proposalIndex === -1) return;
+
+    super.remove(thread);
+    this._threads.splice(proposalIndex, 1);
+    return this;
+  }
+
+  public clear() {
+    super.clear();
+    this._threads = [];
+  }
+
+  // Getters
+
+  public getById(id) {
+    return this._threads.find((t) => t.id === id);
+  }
+
+  public getThreads(params: IListingParams): OffchainThread[] {
+    const { topicName, stageName, pinned } = params;
+
+    let unsortedThreads;
+    if (topicName) {
+      unsortedThreads = this._threads.filter(
+        (t) => t.topic?.name === topicName && t.pinned === pinned
+      );
+    } else if (stageName) {
+      unsortedThreads = this._threads.filter(
+        (t) => t.stage === stageName && t.pinned === pinned
+      );
+    } else {
+      unsortedThreads = this._threads.filter((t) => t.pinned === pinned);
+    }
+
+    return unsortedThreads
+      .filter((t) => this._isBeforeCutoff(params, t))
+      .sort(orderDiscussionsbyLastComment);
+  }
+
+  public getPinnedThreads(): OffchainThread[] {
+    return this._threads
+      .filter((t) => t.pinned)
+      .sort(orderDiscussionsbyLastComment);
+  }
+
+  // Initialization, i.e. "Have we begun fetching threads of this type?"
+
+  public isInitialized(params: IListingParams): boolean {
```

Comment:
Since the params (topic/stage/pinned/none) are mutually exclusive functionally speaking, we may want to make `IListingParams` an enum or union type (e.g. `Topic | Stage | Pinned | All`) rather than an object with optional values.

Then, we can internally reduce the size of these arrays by reworking the types to be something like:

```typescript
type ICategoryState = {
  initialized: boolean;
  depleted: boolean;
  cutoffDate: Moment;
};

const topics: { [id:string]: ICategoryState };
const stages: { [id:string]: ICategoryState };
const allThreads: ICategoryState;
```

This may also let us simplify some of the if/elseif/else branches.

Reply:
So, a couple complications:

1) The params aren't actually mutually exclusive, just topic/stage. Pinned is additive. 

2) I guess I'm a lil unclear how an enum would help, given we need both the topic name, and the fact that it's a topic name (because at the moment, we don't prevent overlap in topic names and stage names). If we ensured mutual exclusivity between stage & topic names, we could have a single initialized[subpage], and no need for distinguishing, for the if/else structures, etc.

3) I did actually swap, at one point, to the ICategoryState structure (initialized, depleted, cutoffDate) you suggest, but my issue became that I needed a check if the relevant topic/stage was initialized at each point. E.g., right now I can treat `isDepleted[topicName]` as false if it's undefined. But if I switch to `listingStore[topicName][isDepleted]`, I now have to check each time to ensure that `listingStore[topicName]` each time I check for `isDepleted` (and I have to define it if it isn't). This (I think) ends up adding more code in the methods, and more fragility to checks, than it adds in concision/clarity to the FetchState interface.

Reply:
1 makes sense.

Re: 2, that also makes sense. Technically the correct way to do it then would be a union type, like:
```typescript
type IListingParams = { topic: string} | { stage: string } | { pinned: boolean };
```
And then using typeguards to distinguish between the three objects, but it's fine as it is.

Re: 3, you can probably get away with `?` notation to avoid that check, so you can do like `listingStore[topicName]?[isDepleted]` and it'll properly avoid throwing if undefined.